ui/app/adapters/job.js

```diff
@@ -159,4 +167,138 @@ export default class JobAdapter extends WatchableNamespaceIDs {
+    const region = this.system.activeRegion;
+    const applicationAdapter = getOwner(this).lookup('adapter:application');
+    const prefix = `${
+      applicationAdapter.host || window.location.host
+    }/${applicationAdapter.urlPrefix()}`;
+
+    const wsUrl =
+      `${protocol}//${prefix}/job/${encodeURIComponent(job.get('id'))}/action` +
+      `?namespace=${job.get('namespace.id')}&action=${
+        action.name
+      }&allocID=${allocID}&task=${action.task.name}&group=${
+        action.task.taskGroup.name
+      }&tty=true&ws_handshake=true` +
+      (region ? `&region=${region}` : '');
+
+    // const socket = new WebSocket(wsUrl);
+    let socket;
+
+    const mirageEnabled =
+      config.environment !== 'production' &&
+      config['ember-cli-mirage'] &&
+      config['ember-cli-mirage'].enabled !== false;
+
+    if (mirageEnabled) {
+      socket = new Object({
+        messageDisplayed: false,
+        addEventListener: function (event, callback) {
+          if (event === 'message') {
+            this.onmessage = callback;
+          }
+          if (event === 'open') {
+            this.onopen = callback;
+          }
+          if (event === 'close') {
+            this.onclose = callback;
+          }
+        },
+
+        send(e) {
+          if (!this.messageDisplayed) {
+            this.messageDisplayed = true;
+            this.onmessage({
+              data: `{"stdout":{"data":"${btoa('Message Received')}"}}`,
+            });
+          } else {
+            this.onmessage({ data: e.replace('stdin', 'stdout') });
+          }
+        },
+      });
+    } else {
+      socket = new WebSocket(wsUrl);
+    }
+
+    let notification;
+    socket.addEventListener('open', () => {
+      notification = this.notifications
+        .add({
+          title: `Action ${action.name} Started`,
+          color: 'neutral',
+          code: true,
+          sticky: true,
+          customAction: {
+            label: 'Stop Action',
+            action: () => {
+              socket.close();
+            },
+          },
+        })
+        .getFlashObject();
+
+      socket.send(
+        JSON.stringify({ version: 1, auth_token: this.token?.secret || '' })
+      );
+      socket.send(
+        JSON.stringify({
+          tty_size: { width: 250, height: 100 }, // Magic numbers, but they pass the eye test.
+        })
+      );
+    });
+
+    socket.addEventListener('message', (event) => {
+      if (!this.notifications.queue.includes(notification)) {
+        // User has manually closed the notification;
+        // explicitly close the socket and return;
+        socket.close();
+        return;
+      }
+
+      let jsonData = JSON.parse(event.data);
+      if (jsonData.stdout && jsonData.stdout.data) {
+        messageBuffer += base64DecodeString(jsonData.stdout.data);
+        messageBuffer += '\n';
+        messageBuffer = messageBuffer.replace(/\x1b\[[0-9;]*[a-zA-Z]/g, '');
```

Comment:
I think this magic could use a comment. Is this some control sequence that's getting noped out? What's the control sequence?

Reply:
These strip out ansi escape chars (`\03`, `\x1b[` etc.) that indicate anything from carriage return to colour change.

I'm treating this less like an interactive exec window and more like a raw output display, so have opted to escape and remove these characters.